File: bst.java
import java.util.*;
import java.io.*;

// ==========================================
// BST DENGAN INPUT FILE CSV
// ==========================================
public class bst {

    // ===== NODE =====
    static class Node {
        int id;
        String nama;
        Node left, right;

        Node(int id, String nama) {
            this.id = id;
            this.nama = nama;
        }
    }

    // ===== BST =====
    static class BST {
        Node root;

        // INSERT
        Node insert(Node node, int id, String nama) {
            if (node == null) return new Node(id, nama);

            if (id < node.id)
                node.left = insert(node.left, id, nama);
            else if (id > node.id)
                node.right = insert(node.right, id, nama);

            return node;
        }

        // LOAD CSV
        void loadCSV(String fileName) {
            try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
                String line;

                br.readLine(); // skip header

                while ((line = br.readLine()) != null) {
                    String[] data;

                    // support ; atau ,
                    if (line.contains(";"))
                        data = line.split(";");
                    else
                        data = line.split(",");

                    int id = Integer.parseInt(data[0].trim());
                    String nama = data[1].trim();

                    root = insert(root, id, nama);
                }

                System.out.println("Semua data berhasil dimasukkan ke BST!");

            } catch (Exception e) {
                System.out.println("❌ Error: " + e.getMessage());
            }
        }

        // SEARCH
        Node search(Node node, int id) {
            if (node == null || node.id == id) return node;
            return (id < node.id) ? search(node.left, id) : search(node.right, id);
        }

        // DELETE
        Node min(Node node) {
            while (node.left != null) node = node.left;
            return node;
        }

        Node delete(Node node, int id) {
            if (node == null) return null;

            if (id < node.id)
                node.left = delete(node.left, id);
            else if (id > node.id)
                node.right = delete(node.right, id);
            else {
                if (node.left == null) return node.right;
                if (node.right == null) return node.left;

                Node temp = min(node.right);
                node.id = temp.id;
                node.nama = temp.nama;
                node.right = delete(node.right, temp.id);
            }
            return node;
        }

        // TRAVERSAL (rapi)
        void inorder(Node node, List<String> list) {
            if (node != null) {
                inorder(node.left, list);
                list.add(node.id + " - " + node.nama);
                inorder(node.right, list);
            }
        }

        void preorder(Node node, List<String> list) {
            if (node != null) {
                list.add(node.id + " - " + node.nama);
                preorder(node.left, list);
                preorder(node.right, list);
            }
        }

        void postorder(Node node, List<String> list) {
            if (node != null) {
                postorder(node.left, list);
                postorder(node.right, list);
                list.add(node.id + " - " + node.nama);
            }
        }

        void printGrid(List<String> list) {
            int col = 3;
            for (int i = 0; i < list.size(); i++) {
                System.out.printf("%-30s", list.get(i));
                if ((i + 1) % col == 0) System.out.println();
            }
            System.out.println();
        }

        void showTraversal() {
            List<String> data = new ArrayList<>();

            System.out.println("\n=== INORDER ===");
            inorder(root, data);
            printGrid(data);

            System.out.println("\n=== PREORDER ===");
            data.clear();
            preorder(root, data);
            printGrid(data);

            System.out.println("\n=== POSTORDER ===");
            data.clear();
            postorder(root, data);
            printGrid(data);
        }
    }

    // ===== MAIN =====
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        BST tree = new BST();

        int pilih;

        do {
            System.out.println("\n=== MENU BST (CSV) ===");
            System.out.println("1. Load Data dari CSV");
            System.out.println("2. Cari Data");
            System.out.println("3. Hapus Data");
            System.out.println("4. Tampilkan Traversal");
            System.out.println("0. Keluar");
            System.out.print("Pilih: ");
            pilih = sc.nextInt();

            switch (pilih) {
                case 1:
                    tree.loadCSV("data100.csv");
                    break;

                case 2:
                    System.out.print("Masukkan ID: ");
                    int cari = sc.nextInt();
                    Node hasil = tree.search(tree.root, cari);

                    if (hasil != null)
                        System.out.println("✔ Ditemukan: " + hasil.id + " - " + hasil.nama);
                    else
                        System.out.println("✖ Tidak ditemukan");
                    break;

                case 3:
                    System.out.print("Masukkan ID: ");
                    int hapus = sc.nextInt();
                    tree.root = tree.delete(tree.root, hapus);
                    System.out.println("✔ Data dihapus");
                    break;

                case 4:
                    tree.showTraversal();
                    break;

                case 0:
                    System.out.println("Program selesai.");
                    break;

                default:
                    System.out.println("Pilihan salah!");
            }

        } while (pilih != 0);

        sc.close();
    }
}

File: bst.py
import csv

# ===== NODE =====
class Node:
    def __init__(self, id, nama):
        self.id = id
        self.nama = nama
        self.left = None
        self.right = None

# ===== BST =====
class BST:
    def __init__(self):
        self.root = None

    # INSERT
    def insert(self, root, id, nama):
        if root is None:
            return Node(id, nama)

        if id < root.id:
            root.left = self.insert(root.left, id, nama)
        elif id > root.id:
            root.right = self.insert(root.right, id, nama)

        return root

    # LOAD CSV
    def load_csv(self, filename):
        try:
            with open(filename, newline='', encoding='utf-8') as file:
                reader = csv.reader(file)

                next(reader)  # skip header

                for row in reader:
                    if len(row) < 2:
                        continue

                    # support ; atau ,
                    if ";" in row[0]:
                        row = row[0].split(";")

                    id = int(row[0].strip())
                    nama = row[1].strip()

                    self.root = self.insert(self.root, id, nama)

            print("✅ Data berhasil dimuat ke BST!")

        except Exception as e:
            print("❌ Error:", e)

    # SEARCH
    def search(self, root, id):
        if root is None or root.id == id:
            return root
        if id < root.id:
            return self.search(root.left, id)
        return self.search(root.right, id)

    # DELETE
    def min_value(self, root):
        while root.left:
            root = root.left
        return root

    def delete(self, root, id):
        if root is None:
            return root

        if id < root.id:
            root.left = self.delete(root.left, id)
        elif id > root.id:
            root.right = self.delete(root.right, id)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left

            temp = self.min_value(root.right)
            root.id = temp.id
            root.nama = temp.nama
            root.right = self.delete(root.right, temp.id)

        return root

    # TRAVERSAL
    def inorder(self, root, result):
        if root:
            self.inorder(root.left, result)
            result.append(f"{root.id} - {root.nama}")
            self.inorder(root.right, result)

    def preorder(self, root, result):
        if root:
            result.append(f"{root.id} - {root.nama}")
            self.preorder(root.left, result)
            self.preorder(root.right, result)

    def postorder(self, root, result):
        if root:
            self.postorder(root.left, result)
            self.postorder(root.right, result)
            result.append(f"{root.id} - {root.nama}")

    # PRINT GRID
    def print_grid(self, data):
        col = 3
        for i, val in enumerate(data):
            print(f"{val:<30}", end="")
            if (i + 1) % col == 0:
                print()
        print()

    def show_traversal(self):
        data = []

        print("\n=== INORDER ===")
        self.inorder(self.root, data)
        self.print_grid(data)

        print("\n=== PREORDER ===")
        data.clear()
        self.preorder(self.root, data)
        self.print_grid(data)

        print("\n=== POSTORDER ===")
        data.clear()
        self.postorder(self.root, data)
        self.print_grid(data)


# ===== MAIN =====
bst = BST()

while True:
    print("\n=== MENU BST (PYTHON) ===")
    print("1. Load Data dari CSV")
    print("2. Cari Data")
    print("3. Hapus Data")
    print("4. Tampilkan Traversal")
    print("0. Keluar")

    pilih = input("Pilih: ")

    if pilih == "1":
        bst.load_csv("D:/Kuliah/Semester 4/Struktur Data/Praktikum/#6 BST/data100.csv")

    elif pilih == "2":
        id = int(input("Masukkan ID: "))
        hasil = bst.search(bst.root, id)

        if hasil:
            print(f"✔ Ditemukan: {hasil.id} - {hasil.nama}")
        else:
            print("✖ Tidak ditemukan")

    elif pilih == "3":
        id = int(input("Masukkan ID: "))
        bst.root = bst.delete(bst.root, id)
        print("✔ Data dihapus")

    elif pilih == "4":
        bst.show_traversal()

    elif pilih == "0":
        print("Program selesai.")
        break

    else:
        print("Pilihan tidak valid!")